### ISSUE TITLE
[Bug]: Speculative decoding cannot match tokens

### ISSUE THREAD

**Issue by jiqing-feng** [vllm-project/vllm#6285]
### Your current environment

Note: The vLLM is built from source by the latest main branch.

```text
Collecting environment information...
PyTorch version: 2.3.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.27.6
Libc version: glibc-2.35

Python version: 3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-4.18.0-425.3.1.el8.x86_64-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.2.140
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA A100 80GB PCIe
GPU 1: NVIDIA A100 80GB PCIe
GPU 2: NVIDIA A100 80GB PCIe
GPU 3: NVIDIA A100 80GB PCIe
GPU 4: NVIDIA A100 80GB PCIe
GPU 5: NVIDIA A100 80GB PCIe
GPU 6: NVIDIA A100 80GB PCIe

Nvidia driver version: 535.54.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.5
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 57 bits virtual
Byte Order:                      Little Endian
CPU(s):                          128
On-line CPU(s) list:             0-127
Vendor ID:                       GenuineIntel
BIOS Vendor ID:                  Intel(R) Corporation
Model name:                      Intel(R) Xeon(R) Platinum 8358 CPU @ 2.60GHz
BIOS Model name:                 Intel(R) Xeon(R) Platinum 8358 CPU @ 2.60GHz
CPU family:                      6
Model:                           106
Thread(s) per core:              2
Core(s) per socket:              32
Socket(s):                       2
Stepping:                        6
CPU max MHz:                     3400.0000
CPU min MHz:                     800.0000
BogoMIPS:                        5200.00
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_p
erfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer
aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust
bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mb
m_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear
pconfig flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       3 MiB (64 instances)
L1i cache:                       2 MiB (64 instances)
L2 cache:                        80 MiB (64 instances)
L3 cache:                        96 MiB (2 instances)
NUMA node(s):                    2
NUMA node0 CPU(s):               0-31,64-95
NUMA node1 CPU(s):               32-63,96-127
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] numpy==1.22.2
[pip3] nvidia-nccl-cu12==2.20.5
[pip3] onnx==1.14.0
[pip3] pytorch-quantization==2.1.2
[pip3] torch==2.3.0
[pip3] torch-tensorrt==0.0.0
[pip3] torchdata==0.7.0a0
[pip3] torchtext==0.16.0a0
[pip3] torchvision==0.18.0
[pip3] transformers==4.42.3
[pip3] triton==2.3.0
[pip3] vllm-nccl-cu12==2.18.1.0.4.0
[conda] Could not collect
ROCM Version: Could not collect
Neuron SDK Version: N/A
vLLM Version: 0.5.1+cu122
vLLM Build Flags:
CUDA Archs: 5.2 6.0 6.1 7.0 7.2 7.5 8.0 8.6 8.7 9.0+PTX; ROCm: Disabled; Neuron: Disabled
GPU Topology:
GPU0    GPU1    GPU2    GPU3    GPU4    GPU5    GPU6    CPU Affinity    NUMA Affinity   GPU NUMA ID
GPU0     X      PIX     PIX     PIX     SYS     SYS     SYS     0-31,64-95      0               N/A
GPU1    PIX      X      PIX     PIX     SYS     SYS     SYS     0-31,64-95      0               N/A
GPU2    PIX     PIX      X      PIX     SYS     SYS     SYS     0-31,64-95      0               N/A
GPU3    PIX     PIX     PIX      X      SYS     SYS     SYS     0-31,64-95      0               N/A
GPU4    SYS     SYS     SYS     SYS      X      PIX     PIX     32-63,96-127    1               N/A
GPU5    SYS     SYS     SYS     SYS     PIX      X      PIX     32-63,96-127    1               N/A
GPU6    SYS     SYS     SYS     SYS     PIX     PIX      X      32-63,96-127    1               N/A

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks
```

### 🐛 Describe the bug

You can see the matched tokens are always 0 by this PR #6284 

```python
from vllm.entrypoints.llm import LLM
from vllm.sampling_params import SamplingParams


model = "meta-llama/Llama-2-7b-chat-hf"
assistant_model = "meta-llama/Llama-2-7b-chat-hf"

sampling_params = SamplingParams(top_k=1, max_tokens=32, min_tokens=32)
llm = LLM(
        model=model,
        speculative_model=assistant_model, # The draft model. Must have same vocabulary as target model.
        tensor_parallel_size=1,
        num_speculative_tokens=1, # The number of speculative tokens to score.
        use_v2_block_manager=True,
        max_num_seqs=1,
        enforce_eager=True,
    )

prompt = ["This is an example input"]
llm.generate(prompt, sampling_params)
```

When you run this code, you can see the number of match tokens is always 0.
```
predict tokens: 1
matched tokens: 0
predict tokens: 1
matched tokens: 0
predict tokens: 1
matched tokens: 0
predict tokens: 1
matched tokens: 0
predict tokens: 1
matched tokens: 0
predict tokens: 1
matched tokens: 0
predict tokens: 1
matched tokens: 0
predict tokens: 1
matched tokens: 0
predict tokens: 1
matched tokens: 0
predict tokens: 1
matched tokens: 0
predict tokens: 1
matched tokens: 0
predict tokens: 1
matched tokens: 0
predict tokens: 1
matched tokens: 0
............
```
(repeat 32 times, which means the target model runs 32 times.)

You can change the prompt and the sampling_params to verify it.

When I change `num_speculative_tokens=2`, the outputs will be:
```
predict tokens: 2
matched tokens: 1
predict tokens: 2
matched tokens: 1
predict tokens: 2
matched tokens: 1
predict tokens: 2
matched tokens: 1
predict tokens: 2
matched tokens: 1
predict tokens: 2
matched tokens: 1
predict tokens: 2
matched tokens: 1
predict tokens: 2
matched tokens: 1
predict tokens: 2
matched tokens: 1
predict tokens: 2
matched tokens: 1
predict tokens: 2
matched tokens: 1
predict tokens: 2
matched tokens: 1
predict tokens: 2
matched tokens: 1
predict tokens: 2
matched tokens: 1
predict tokens: 2
matched tokens: 1
predict tokens: 2
matched tokens: 1
```
(repeat 16 times)

You can see the number of matched tokens is always 1.

**Comment by jiqing-feng** [comment]
Hi @cadedaniel . It seems like a critical bug of speculative decoding; I am looking into it, but I am not sure which part caused the issue. Would you please take a look at it? Thx!

**Comment by jiqing-feng** [comment]
Fixed by #5765
